File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2024/Y24D20.kt
package org.gristle.pdxpuzzles.advent.y2024

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.iteration.minMax
import org.gristle.pdxpuzzles.utilities.objects.Coord
import org.gristle.pdxpuzzles.utilities.objects.toCoord
import java.util.ArrayDeque
import kotlin.math.absoluteValue
import kotlin.math.max
import kotlin.math.min

class Y24D20(private val racetrack: String) : Day {
    private val width = racetrack.indexOf('\n') + 1
    private val height = (racetrack.length + 1) / width
    private val movements = listOf(-width, 1, width, -1)
    private val start = racetrack.indexOf('S')
    private val end = racetrack.indexOf('E')
    private val visited = IntArray(racetrack.length) { 1_000_000 }
    private val fromStart = bfs(start, end, visited.clone().also { it[start] = 0 }, ::getNeighbors)
    private val fromEnd = bfs(end, start, visited.also { it[end] = 0 }, ::getNeighbors)
    private val honestDistance = fromStart[end]
    private val threshold = honestDistance - 100

    private inline fun <T> bfs(start: Int, end: Int?, visited: T, edges: (Int, T, Int) -> List<Pair<Int, Int>>): T {
        val q = ArrayDeque<Pair<Int, Int>>()
        q.addLast(0 to start)

        while (q.isNotEmpty()) {
            val (steps, current) = q.removeFirst()
            if (current == end) break
            val neighbors = edges(current, visited, steps)
            q.addAll(neighbors)
        }
        return visited
    }

    private fun getNeighbors(
        current: Int,
        visited: IntArray,
        steps: Int
    ): List<Pair<Int, Int>> = movements.mapNotNull { movement ->
        val neighbor = current + movement
        when {
            visited[neighbor] != 1_000_000 -> null
            racetrack[neighbor] == '#' -> {
                visited[neighbor] = steps + 1
                null
            }

            else -> {
                val neighborSteps = steps + 1
                visited[neighbor] = neighborSteps
                neighborSteps to neighbor
            }
        }
    }

    private fun countValid(
        pos: Int,
        endCoord: Coord,
        steps: Int,
    ): Int {
        if (racetrack[pos] == '#') return 0

        val posCoord = pos.toCoord(width)
        val difference = endCoord - posCoord
        val minSteps = steps + difference.manhattanDistance()

        if (minSteps > threshold) return 0

        var count = 0
        val allowance = (threshold - minSteps) / 2
        val (north, south) = minMax(posCoord.y, endCoord.y)
        val yMin = min(north - allowance, posCoord.y - 20).coerceAtLeast(1)
        val yMax = min(south + allowance, posCoord.y + 20).coerceAtMost(height - 2)
        val yRange = yMin..yMax
        for (y in yRange) {
            val (west, east) = minMax(posCoord.x, endCoord.x)
            val xAllowance = when {
                y < north -> allowance - (north - y)
                y > south -> allowance - (y - south)
                else -> allowance
            }
            val totalAllowed = 20 - (posCoord.y - y).absoluteValue
            val xMin = max(posCoord.x - totalAllowed, west - xAllowance).coerceAtLeast(1)
            val xMax = min(posCoord.x + totalAllowed, east + xAllowance).coerceAtMost(width - 3)
            val xRange = xMin..xMax
            for (x in xRange) {
                val rePos = y * width + x
                if (racetrack[rePos] == '#') continue
                val cheatSteps = steps + posCoord.manhattanDistance(Coord(x, y)) + fromEnd[rePos]
                if (cheatSteps <= threshold) count++
            }
        }
        return count
    }

    override fun part1(): Int = fromStart.withIndex().count { (pos, steps) ->
        racetrack[pos] == '#' && steps + fromEnd[pos] <= threshold
    }

    override fun part2() = fromStart.withIndex().sumOf { (pos, steps) ->
        countValid(pos, end.toCoord(width), steps)
    }
}

fun main() = Day.runDay(Y24D20::class)

//    Class creation: 9ms
//    Part 1: 1406 (4ms)
//    Part 2: 1006101 (90ms)
//    Total time: 104ms

@Suppress("unused")
private val test = listOf("""###############
#...#...#.....#
#.#.#.#.#.###.#
#S#...#.#.#...#
#######.#.#.###
#######.#.#...#
#######.#.###.#
###..E#...#...#
###.#######.###
#...###...#...#
#.#####.#.###.#
#.#...#.#.#...#
#.#.#.#.#.#.###
#...#...#...###
###############""")

File: rust/advent/src/bin/2024/y24d20.rs
use std::{cmp::{max, min}, collections::VecDeque};

use advent::utilities::get_input::get_input;
use itertools::Itertools;
use utilities::structs::{coord::Coord2, stopwatch::{ReportDuration, Stopwatch}, str_grid::{AdjacentMetadata, StrGrid}};

type Output = usize;
type Pos = Coord2;

#[derive(Debug)]
struct Input<'a> {
    racetrack: StrGrid<'a>,
    end: usize,
    from_start: Vec<usize>,
    from_end: Vec<usize>,
    threshold: usize,
}

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(24, 20).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn parse_input(input: &str) -> Input {
    let racetrack = StrGrid::new(input).unwrap();
    let start = input.find('S').unwrap();
    let end  = input.find('E').unwrap();
    let mut visited_from_start: Vec<usize> = vec![1_000_000; input.len()];
    visited_from_start[start] = 0;
    let mut visited_from_end = vec![1_000_000; input.len()];
    visited_from_end[end] = 0;
    let from_start = bfs(&racetrack, start, end, visited_from_start);
    let from_end = bfs(&racetrack, end, start, visited_from_end);
    let threshold = from_start[end] - 100;
    Input {
        racetrack,
        end,
        from_start,
        from_end,
        threshold,
    }
}

fn bfs(
    racetrack: &StrGrid,
    start: usize, 
    end: usize, 
    mut visited: Vec<usize>, 
)-> Vec<usize> {
    let mut q = VecDeque::new();
    q.push_back((0, start));
    while let Some((steps, current)) = q.pop_front() {
        if current == end { 
            break; 
        }
        let neighbors = racetrack.adjacent(current)
            .filter_map(|AdjacentMetadata { pos, dir: _, b }| {
                if visited[pos] != 1_000_000 { return None; }
                if b == b'#' {
                    visited[pos] = steps + 1;
                    None
                } else {
                    let neighbor_steps = steps + 1;
                    visited[pos] = neighbor_steps;
                    Some((neighbor_steps, pos))
                }
            });
        for neighbor in neighbors {
            q.push_back(neighbor);
        }
    }
    visited
}

fn count_valid(input: &Input, pos: usize, end_coord: Pos, steps: usize) -> usize {
    if input.racetrack.get(pos) == Some(b'#') { return 0; }

    let pos_coord = Pos::from_index(pos, input.racetrack.width).unwrap();
    let difference = end_coord - pos_coord;
    let min_steps = steps + difference.manhattan_distance(&Pos::origin());
    
    if min_steps > input.threshold { return 0; }
    
    let mut count = 0;
    let allowance = ((input.threshold - min_steps) / 2) as i64;
    let (north, south) = [pos_coord.y(), end_coord.y()].into_iter()
        .minmax()
        .into_option()
        .unwrap();
    let y_min = max(1, min(north - allowance, pos_coord.y() - 20));
    let y_max = min(
        (input.racetrack.height - 2) as i64, 
        min(south + allowance, pos_coord.y() + 20)
    );
    for y in y_min..=y_max {
        let (west, east) = [pos_coord.x(), end_coord.x()].into_iter()
            .minmax()
            .into_option()
            .unwrap();
        
        let x_allowance = if y < north {
            allowance - (north - y)
        } else if y > south {
            allowance - (y - south)
        } else {
            allowance
        };
        
        let total_allowed = 20 - (pos_coord.y() - y).abs();
        let x_min = max(
            1,
            max(pos_coord.x() - total_allowed, west - x_allowance)
        );
        let x_max = min(
            (input.racetrack.width - 3) as i64,
            min(pos_coord.x() + total_allowed, east + x_allowance)
        );
        for x in x_min..=x_max {
            let re_pos = (y * input.racetrack.width as i64 + x) as usize;
            if input.racetrack.get(re_pos) == Some(b'#') { continue; }
            let cheat_steps = steps 
                + pos_coord.manhattan_distance(&Pos::new2d(x, y))
                + input.from_end[re_pos];
            if cheat_steps <= input.threshold {
                count += 1;
            }
        }
    }
    count
}


fn part1(input: &Input) -> Output {
    input.from_start.iter().enumerate()
        .filter(|&(pos, &steps)| {
            input.racetrack.get(pos) == Some(b'#') 
                && steps + input.from_end[pos] <= input.threshold
        })
        .count()
}

fn part2(input: &Input) -> Output {
    input.from_start.iter().enumerate()
        .map(|(pos, &steps)| {
            count_valid(
                input,
                pos, 
                Pos::from_index(input.end, input.racetrack.width).unwrap(), 
                steps
            )
        })
        .sum()
}

#[test]
fn default() {
    let input = get_input(24, 20).unwrap();
    let input = parse_input(&input);
    assert_eq!(1406, part1(&input));
    assert_eq!(1006101, part2(&input));
}

// Input parsed (397μs)
// 1. 1406 (35μs)
// 2. 1006101 (16ms)
// Total: 16ms